File: collatz.erl
%
% Collatz sequence generator. Erlang version.
% (http://en.wikipedia.org/wiki/Collatz_conjecture)
%
% Copyright (c)2011, Ivan Tam. ivan@hipnik.net
%

-module(collatz).
-export([sequence/1]).

sequence( X ) -> sequence(X, []).

sequence( 1, _ )     -> [1];
sequence( X, [] )    -> sequence(X, [X]);
sequence( X, [H|T] ) ->
  case (X rem 2) of
    0 -> [ H | sequence( X div 2,   T) ];
    1 -> [ H | sequence( X * 3 + 1, T) ]
  end.

File: python/collatz.py
#
# Collatz sequence generator.
# (http://en.wikipedia.org/wiki/Collatz_conjecture)
#
# Recursive (with accumulator) Python version.
#
# Copyright (c)2011, Ivan Tam. ivan@hipnik.net
#

def collatz(n, acc=None):
  # if the accumulator is not set, give it a head
  if acc is None: acc = [n]

  # base case
  if n == 1: return [1]

  if n % 2 == 1:     # odd case
    next = n * 3 + 1
  elif n % 2 == 0:   # even case
    next = n / 2

  acc.append( next )
  collatz( next, acc )
  return acc

if __name__ == '__main__':
  print collatz(1024)
  print collatz(50)
  print collatz(1)
